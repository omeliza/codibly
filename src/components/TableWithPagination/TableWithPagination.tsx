import { Key, useState } from 'react';
import TableCell from "@mui/material/TableCell";
import { Paper, Table, TableBody, TableContainer, TableHead, TablePagination, TableRow } from '@mui/material';
import { useAppSelector } from '../../hooks/hooks';
import { StyledTableCell } from './styles';

interface Column {
  id: "id" | "name" | "year";
  label: string;
  minWidth?: number;
  align?: "right";
  format?: (value: number) => string;
}

const columns: readonly Column[] = [
  { id: "id", label: "ID", minWidth: 50 },
  { id: "name", label: "Name", minWidth: 120 },
  {
    id: "year",
    label: "Year",
    minWidth: 80,
    align: "right",
  }
];


export const TableWithPagination = () => {
  const [page, setPage] = useState(0);
  const [rowsPerPage, setRowsPerPage] = useState(5);

  const { colors, isLoading, error, searchText } = useAppSelector(state => state.colorReducer);

  const handleChangePage = (event: unknown, newPage: number) => {
    setPage(newPage);
  };

  const handleChangeRowsPerPage = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    setRowsPerPage(+event.target.value);
    setPage(0);
  };

  const filteredColors = (value: number | undefined) => {
    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
    return value
      ? colors.filter((item: { id: { toString: () => string | string[]; }; }) => item.id.toString().includes(value.toString()))
      : colors;
  }

  return (
    <>
      <Paper sx={{ maxWidth: "1000px", overflow: "hidden", display: 'flex', flexDirection: 'column', justifyContent: 'center', m:'0 auto' }}>
        <TableContainer sx={{ maxHeight: 440 }}>
          <Table stickyHeader aria-label="sticky table">
            <TableHead>
              <TableRow>
                {columns.map((column) => (
                  <StyledTableCell
                    key={column.id}
                    align={column.align}
                    style={{ minWidth: column.minWidth }}
                  >
                    {column.label}
                  </StyledTableCell>
                ))}
              </TableRow>
            </TableHead>
            <TableBody>
              {filteredColors(searchText)
                .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
                .map((row: { [x: string]: any; id: Key | null | undefined; color: any; }) => {
                  return (
                    <TableRow hover tabIndex={-1} key={row.id} style={{ backgroundColor: `${row.color}`}}>
                      {columns.map(column => {
                        const value = row[column.id];
                        return (
                          <TableCell key={column.id} align={column.align}>
                            {column.format && typeof value === "number"
                              ? column.format(value)
                              : value}
                          </TableCell>
                        );
                      })}
                    </TableRow>
                  )
                })}
            </TableBody>
          </Table>
        </TableContainer>
        <TablePagination
          rowsPerPageOptions={[5]}
          component="div"
          count={filteredColors(searchText).length}
          rowsPerPage={rowsPerPage}
          page={page}
          onPageChange={handleChangePage}
          onRowsPerPageChange={handleChangeRowsPerPage}
        />
      </Paper>
      {isLoading && <h1>Loading...</h1>}
      {error && <h1>{error}</h1>}
    </>
  )
}

