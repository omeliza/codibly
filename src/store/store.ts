import { combineReducers, configureStore } from "@reduxjs/toolkit";
// import { colorAPI } from "../services/DataService";
import colorReducer from './reducers/ColorSlice';

const rootReducer = combineReducers({
  colorReducer,
  // [colorAPI.reducerPath]: colorAPI.reducer
})

export const setupStore = () => {
  return configureStore({
    reducer: rootReducer,
    // middleware: (getDefaultMiddleware) =>
    //   getDefaultMiddleware().concat(colorAPI.middleware)
  })
}

export type RootState = ReturnType<typeof rootReducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']