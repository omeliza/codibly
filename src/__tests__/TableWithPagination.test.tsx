import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { setupStore } from '../store/store';

import { TableWithPagination } from '../components/TableWithPagination/TableWithPagination';

const store = setupStore();

describe('TableWithPagination', () => {
  it('should render correctly', () => {
    const tree = renderer.create(
    <Provider store={store}>
      <TableWithPagination />
    </Provider>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});