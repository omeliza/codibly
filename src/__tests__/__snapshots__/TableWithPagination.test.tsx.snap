// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableWithPagination should render correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-9x6fsg-MuiPaper-root"
>
  <div
    className="MuiTableContainer-root css-41abqd-MuiTableContainer-root"
  >
    <table
      aria-label="sticky table"
      className="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
      role={null}
    >
      <thead
        className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-67bhig-MuiTableCell-root"
            scope="col"
            style={
              Object {
                "minWidth": 50,
              }
            }
          >
            ID
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-67bhig-MuiTableCell-root"
            scope="col"
            style={
              Object {
                "minWidth": 120,
              }
            }
          >
            Name
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignRight MuiTableCell-sizeMedium css-s97c5p-MuiTableCell-root"
            scope="col"
            style={
              Object {
                "minWidth": 80,
              }
            }
          >
            Year
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        role={null}
      />
    </table>
  </div>
  <div
    className="MuiTablePagination-root css-jtlhu6-MuiTablePagination-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
    >
      <div
        className="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
      />
      <p
        className="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
      >
        0â€“0 of 0
      </p>
      <div
        className="MuiTablePagination-actions"
      >
        <button
          aria-label="Go to previous page"
          className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Go to previous page"
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyboardArrowLeftIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
            />
          </svg>
        </button>
        <button
          aria-label="Go to next page"
          className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Go to next page"
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyboardArrowRightIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
